### ISSUE TITLE
Investigate recreating a GPT-2 forward pass with KerasNLP

### ISSUE THREAD

**Issue by mattdangerw** [keras-team/keras-nlp#337]
With the addition of a roughly compatible tokenizer with gpt-2 in https://github.com/keras-team/keras-nlp/pull/303, it should be possible to start experimenting with recreating an entire gpt forward pass with KerasNLP components.

This should be more an investigation and validation for now, as we continue to answer some core design questions with our BERT modeling code.

Rough flow:

 - Load a reference version of the gpt-2 tokenizer and network, either from the [released github](https://github.com/openai/gpt-2) or huggingface.
 - Recreate the gpt-2 tokenizer using https://github.com/keras-team/keras-nlp/pull/303
 - Recreate the decoder models itself using KerasNLP components, and convert weights.
 - Verify that our output can match with a high degree of accuracy.

**Comment by abheesht17** [comment]
Tried copying the weights present in the official GPT-2 repo here: https://colab.research.google.com/drive/1zjmKuv--nwTbxHM1jeL0hPL-brHJXwCd?usp=sharing. There is a big difference in outputs though.

On further inspection, there seems to be a difference between KerasNLP's `TransformerDecoder` and Hugging Face's GPT-2 decoder.


KerasNLP TransformerDecoder:

https://github.com/keras-team/keras-nlp/blob/b88ff1ba859846c35d9701f5acaf90fecbb7f908/keras_nlp/layers/transformer_decoder.py#L269-L279

1. Apply Attn.
2. Add original input and attention output.
3. Apply LayerNorm to the sum.


Hugging Face:


https://github.com/huggingface/transformers/blob/65fb71bc762c46bb067306c1fd083b1cba87a095/src/transformers/models/gpt2/modeling_gpt2.py#L405-L417

1. Apply LayerNorm.
2. Apply Attn.
3. Add original input and attention output.

We will have to implement a custom GPT-2 decoder layer.

**Comment by abheesht17** [comment]
Discussed offline. Tweaked the `TransformerDecoder` layer. Able to reproduce output: https://colab.research.google.com/drive/1zjmKuv--nwTbxHM1jeL0hPL-brHJXwCd?usp=sharing

**Comment by mattdangerw** [comment]
I think adding a `normalize_first=False` argument to our `TransformerEncoder` and `TransformerDecoder` blocks makes sense. And this is the only change we need for GPT-2 with `TransformerDecoder`, correct?

We want to be careful about making these basic transformer layers "god objects", and generally not venture too far from the *Attn is all you need* formulation here. But this seems like a common and simple enough change to just add a parameter. It seems somewhat silly to have a `keras_nlp.layers.TransformerDecoderWhichNormalizesFirst` class.

@jbischof sound good to you re adding a parameter?

**Comment by jbischof** [comment]
I think this makes sense @mattdangerw! The reasoning is that `TransformerDecoder` is a low-level and heavy-lifting module that would be very hard for users to recreate. The reservations I expressed in #343 were about high level convenience wrappers offering many usage modalities that might confuse new users without offering much to advanced users.

@abheesht17 it would be helpful if you could import your proposed changes from a git branch rather than re-creating all the code in the notebook because it's hard to figure out what's going on at a glance otherwise. Thanks!

**Comment by mattdangerw** [comment]
Actually, since I think we have alignment here... @abheesht17 want to just open a PR to add a parameter `normalize_first=False` to both `TranformerEncoder` and `TransformerDecoder`?

We can discuss any naming concerns on the PR.

**Comment by abheesht17** [comment]
@mattdangerw, @jbischof - sure, will open a PR for `normalize_first` first! Do we also want to add this parameter: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/layers/transformer_encoder_block.py#L124-L126?

**Comment by mattdangerw** [comment]
`diff_q_and_kv_attention_layer_norm`? I would very much hope we do not ever add that to the main TransformerEncoder/TransformerDecoder, we don't even support a separate `key_value` input like the block you are linking. That's the level of customization I would want to expose as a separate layer (either model specific, or generic if we really think it's a general purpose one).

Is there a reason we need that? Is that needed for gpt? If so, I might have an incomplete understanding of this problem.

**Comment by abheesht17** [comment]
Nah, we don't need it for GPT-2, just asked because it is related to `norm_first`. They define a separate LayerNorm layer for the **cross-attention layer**, where they have different QKVs. I guess we can skip it, lel.

**Comment by mattdangerw** [comment]
@abheesht17 with #350 landed, I think we can go ahead and make a "forward pass" PR similar to #304 but for GPT-2.
